### PR TITLE
fix(morphology): anchor the reflected dilation kernel at its origin

### LINKED ISSUE #4733: morphology: dilation reflects the kernel but anchors the window at the unreflected origin, so even kernels and custom origins are off by one pixel
## Summary

`kornia.morphology.dilation` reflects the structuring element (Minkowski dilation, the convention ratified in #4713) but pads the image with the **unreflected** origin, so its window is anchored at `(k_h - 1 - o_row, k_w - 1 - o_col)` instead of at `origin`. For an odd kernel at the default centre the two coincide and the result is right. For every even-sized kernel, and for every non-centre `origin`, the output is shifted by one pixel, and `opening`, `closing`, `gradient`, `top_hat` and `bottom_hat` inherit the shift. `erosion` (no reflection) is anchored correctly.

```python
import torch
from kornia.morphology import dilation, erosion, opening

x = torch.zeros(1, 1, 6, 7); x[..., 2, 3] = 1.0          # one hot pixel, off both centre lines

dilation(x, torch.ones(2, 2))[0, 0].nonzero().tolist()
# [[2, 3], [2, 4], [3, 3], [3, 4]]   kornia
# [[1, 2], [1, 3], [2, 2], [2, 3]]   scipy.ndimage.grey_dilation(footprint=ones((2, 2), bool)) and skimage.morphology.dilation

dilation(x, torch.ones(3, 3), origin=[0, 0])[0, 0].nonzero()[[0, -1]].tolist()
# [[0, 1], [2, 3]]   kornia: the window spreads toward -rows/-cols
# [[2, 3], [4, 5]]   scipy origin=(-1, -1): toward +rows/+cols, the same window erosion(..., origin=[0, 0]) reads

blk = torch.zeros(1, 1, 9, 11); blk[..., 3:6, 3:7] = 1.0
torch.equal(opening(blk, torch.ones(3, 3), origin=[0, 0]), blk)   # False; an opening must leave a block that contains the SE
k = torch.ones(2, 2); xr = torch.rand(1, 1, 9, 11, generator=torch.Generator().manual_seed(7))
torch.equal(erosion(xr, k, origin=[1, 1]), -dilation(-xr, k.flip((0, 1)), origin=[0, 0]))   # False; the adjunction fails
```

Measured on `main` @ 0e8cf3b57, torch 2.14.0, CPU float32; references scipy 1.17.1, scikit-image 0.26.0, opencv-python-headless 5.0.0.

## Cause

`8bbefba71` (#1034, 2021-05) introduced `pad_e = [origin[1], se_w - origin[1] - 1, origin[0], se_h - origin[0] - 1]`; `225160cbd` (#1084, two weeks later) added `neighborhood.flip((0, 1))` (and `.flip(0)` on the convolution bias) to reflect the kernel, without mirroring the pad. Every in-tree test uses an odd kernel at the default origin, where the two cancel, and `test_custom_origin` asserts only that a custom origin changes the output.

## Comparison

| | reflects the kernel | even 2x2 on the hot pixel | `origin`/`anchor` semantics |
|---|---|---|---|
| kornia today | yes | rows {2,3} x cols {3,4} | index, but anchored at the mirrored cell in `dilation` |
| kornia after the fix | yes | rows {1,2} x cols {2,3} | index (author's ruling on #4713) |
| scipy.ndimage `grey_dilation` | yes | rows {1,2} x cols {2,3} | offset from `k // 2` (`origin = index - k // 2`) |
| scikit-image `dilation` | no (mirrors inside `opening`/`closing`) | rows {1,2} x cols {2,3} | none; default = `k // 2` |
| OpenCV `dilate` | no | rows {2,3} x cols {3,4} | `anchor=(x, y)` index, opposite direction because nothing is reflected |
| intcomp/torchmorph `grey_dilation` | yes | rows {1,2} x cols {2,3} (numpy emulation of its CUDA index arithmetic) | scipy offset |

## Expected

With `origin = o`, `dilation(x, B)(p) = max over kernel cells q of x(p - (q - o))`, so `dilation(x, ones(2, 2))` of the hot pixel lands on rows {1,2} x cols {2,3}, `origin=[0, 0]` spreads toward +rows/+cols, `erosion(x, B, o) == -dilation(-x, B.flip((0, 1)), [k_h-1-o_row, k_w-1-o_col])` for every `o`, and `opening` with any origin is anti-extensive and idempotent. The fix is the mirrored pad in `dilation`, `pad_e = [se_w - origin[1] - 1, origin[1], se_h - origin[0] - 1, origin[0]]`; both engines share the padded tensor, so one line covers both. Odd kernels at the default origin are bitwise unchanged. A brute-force Minkowski oracle over six kernels x every origin x two images gives 60 mismatching cases today and 0 after the fix. A PR with that line, pins and a changelog fragment follows (batch 7 of the conventions programme, #4620).

## Classification

Bug (wrong output). Idiom: none of 1-9 fits exactly; the closest is 7 (a slip hidden by symmetric test data: every test kernel is odd and centred, where the reflection and the anchor cancel). Candidate wording, not numbered: *a reflection applied to a kernel's contents but not to its origin, invisible while every fixture is symmetric under that reflection*.

## Disposition

`n/a — single symbol` (the composites delegate to `dilation`; nothing to merge).

Related: #4713 (convention thread), #2466 (asked for the flip to be removed or documented in 2023), #1030 / #1045 / #1062 (historical reports that predate #1034/#1084).

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by ducha-aiki**
Fixes #4733.

## What changes

One line in `kornia.morphology.dilation`: the image is now padded with the **reflected** origin, `pad_e = [se_w - o_col - 1, o_col, se_h - o_row - 1, o_row]`. `dilation` has reflected the kernel contents since #1084 (Minkowski dilation, `out(x) = max_{b∈B} f(x - b)`, the convention the module author confirmed on #4713) but kept the pad of #1034, so its window was anchored at `(k_h - 1 - o_row, k_w - 1 - o_col)` instead of at `origin`. Both engines read the same padded tensor and both flip the neighbourhood, so the one line covers `unfold` and `convolution`. `erosion` is untouched (it never reflected and was anchored correctly).

- Odd kernel at the default origin: bitwise unchanged (the reflected and unreflected centre coincide). Every existing test and doctest in the package uses that case and keeps its literals.
- Even kernel: `dilation(hot(2, 3) in a 6x7, ones(2, 2))` moves from rows {2,3} x cols {3,4} to rows {1,2} x cols {2,3} = `scipy.ndimage.grey_dilation` = `skimage.morphology.dilation` (OpenCV keeps the old set because it does not reflect).
- Custom `origin`: `origin=[0, 0]` on a 3x3 now spreads toward +rows/+cols, the same window `erosion(..., origin=[0, 0])` reads; opening/closing with a custom origin become anti-extensive/extensive and idempotent, and `erosion(x, B, o) == -dilation(-x, B.flip((0, 1)), [k_h-1-o_row, k_w-1-o_col])` holds for every origin.
- `opening`, `closing`, `gradient`, `top_hat`, `bottom_hat` inherit the fix through `dilation`.

## Verification

- Brute-force Minkowski oracle `out(p) = max_{q: kernel[q] != 0} x(p - (q - origin))` over six kernels (`[[0,1,1]]`, L, `ones(2,2)`, `ones(3,3)`, `ones(1,4)`, `ones(4,1)`) x every origin x {hot 6x7, rand 7x10} x both engines: before the fix 60 of 66 (kernel, origin, image) cases mismatch; after it 0 mismatch wherever a kernel member reaches the image. Cross-checked in the batch-7 audit against scipy 1.17.1, scikit-image 0.26.0 and opencv-python-headless 5.0.0: after the fix kornia reproduces scipy bit-for-bit on all six kernels and all 30 origins.
- 7 new pins (`TestDilate` x5, `TestOpening`, `TestClosing`), each RED on the old code and GREEN after: even-kernel centre (square and, on both engines, `ones(1,4)` / `ones(4,1)` with a hot pixel off both centre lines), custom-origin direction, the 13-origin erosion/dilation duality, opening/closing invariants with `origin=[0,0]`, and an odd-kernel/default-origin literal captured from `main` before the edit.
- Mutations: `origin = [(k-1)//2, ...]` turns the even-centre pin red; reverting the pad order turns 5 pins red; swapping the row/col halves of the pad list turns the non-square pin red (8/8 cases) plus 4 duality cases.
- `tests/morphology --dtype=all`: 195 -> 227 passed, 0 failed (+4 dtypes x 8 parametrised pins). MPS float32: 55 passed, 7 pre-existing skips. `tests/filters` (the only other package that calls these functions, through 3x3 default-origin kernels): unchanged. `pre-commit-all` and `typecheck` clean.
- Changelog fragment `changelog.d/4737.fixed.md` says what used to happen and what happens now.

## Not verified here

CUDA and the torch 2.5.1 floor (the change is a pad-order swap; no kernel or API differs across versions). Speed is unchanged (same pad amounts, reordered).

## Context

Batch 7 (morphology) of the conventions programme (#4620). Follow-ups filed from the same audit, not addressed here: #4734 (`max_val` sentinel), #4735 (bool/integer input), #4736 (validation messages). The docstring Convention blocks land in a separate PR stacked on this one. Draft PR #4728 edits the same `dilation` body and will need a trivial rebase.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### CODE REVIEW

**Comment by ducha-aiki** [comment]
Rebased onto `main` after #4728 (engine="auto") merged; head is now `b83844cf5`.

- `morphology.py` merged cleanly. #4728's `"shift"` engine pads the same `pad_e` and reduces over the same `neighborhood.flip((0, 1))`, so the reflected-origin fix covers `"unfold"`, `"convolution"` and `"shift"` alike.
- `119381ae0`: #4728's `test_shift_engine_signed_zero_is_equal_value` used `origin=[0, 0]` on a 2×2 kernel. That covered the whole input only under the old, unreflected anchor. With the anchor fixed, the same window is `origin=[1, 1]`, and it still ties `-0` against `+0`.
- `b83844cf5`: renamed the fragment to `4737.breaking.md`. Even kernels and custom origins now return output shifted by one pixel, so a caller that compensated for the old anchor gets a different result. That is the same bar as #4683. The text already says what used to happen.
- Locally (torch 2.14): `tests/morphology` 1037 passed on CPU in float16/bfloat16/float32/float64, 263 passed on MPS float32, and the dynamo tests pass under `KORNIA_TEST_OPTIMIZER=inductor`.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
